Upload and Download of Prometheus configuration file

rhsakarpos opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:
This PR enables us to 

1. Upload a configuration file to Prometheus
2. Download the configuration file from Prometheus

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
NA

**Special notes for your reviewer**:
The similar API is going to be used to upload and download configuration files for all telemetry 3rd party components, like Prometheus, Alert manager, Node exporter etc. The general flow is as below

*Upload*

- Backup the current configuration file with extension .bak
- Upload the configuration file to the server, in the right location
- Call some API or POST request to reload the configuration (if the 3rd party supports it)

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
**Upload API**
```
http://localhost:50040/v1beta/metrics/uploadconf?conftype=prometheus
// the value of the conftype determines the component, like 'prometheus', 'nodeexporter' etc.
```

**Download API**
```
http://localhost:50040/v1beta/metrics/downloadconf?conftype=prometheus
// the value of the conftype determines the component, like 'prometheus', 'nodeexporter' etc.
```

rhsakarpos commented:
@skdwriting @wisererik @xing-yang Please review

rhsakarpos commented:
This is tested for Prometheus, am now extending it to the other 3rd party components too...that is a WIP

xing-yang commented on pkg/api/routers/fileops.go:
```diff
@@ -0,0 +1,35 @@
+
+			// All operations of file can be used for both admin and users.
+			beego.NSRouter("/uploadconf", controllers.NewFileOpsPortal(), "post:UploadConfFile"),
+			beego.NSRouter("/downloadconf", controllers.NewFileOpsPortal(), "get:DownloadConfFile"),
```
Why do we need APIs for uploading/downloading config files?

rhsakarpos replied:
This was added to add functionality to telemetry after the initial installation. Some use cases are

- add a new exporter (CEPH) to Prometheus 
- define new threshold rules in Prometheus to raise Alerts to Alert manager
etc. 
The primary driver for adding these API was that during a demo (in a meetup, conference etc.) we do NOT need to keep switching between the UI and the backend console to do these configuration changes (either by running a script or by directly editing the configuration files)

leonwanghui replied:
I don't think it's a good idea to add a new first-level API, if this is a required operation, please move it under metrics path.

rhsakarpos replied:
added subpath ```metrics/```

leonwanghui commented on pkg/api/controllers/fileops.go:
```diff
@@ -0,0 +1,204 @@
+)
+
+// prometheus constants
+var PROMETHEUS_CONF_HOME string
```
Please define the variable in this format like `PrometheusCONFHome`, also this style would be better:
```go
var (
    PrometheusCONFHome string
    PrometheusURL string
)
```

rhsakarpos replied:
refactored as per IDE suggestion, to following
var PrometheusConfHome string
var PrometheusUrl string
var PrometheusConfFile string

// alert manager constants
var AlertmgrConfHome string
var AlertmgrUrl string
var AlertmgrConfFile string

var GrafanaConfHome string
var GrafanaRestartCmd string
var GrafanaConfFile string

var ReloadPath string
var BackupExtension string

leonwanghui commented on pkg/api/controllers/fileops.go:
```diff
@@ -0,0 +1,204 @@
+	CtrClient client.Client
+}
+
+func (this *FileOpsPortal) UploadConfFile() {
```
Please don't use `this` as the pointer value, change it to `f`.

rhsakarpos replied:
changed

leonwanghui commented on pkg/api/controllers/fileops.go:
```diff
@@ -0,0 +1,204 @@
+
+	switch confType {
+	case "prometheus":
+		{
```
There is no need to add this big parantheses.

rhsakarpos replied:
removed

leonwanghui commented on pkg/api/controllers/fileops.go:
```diff
@@ -0,0 +1,204 @@
+
+	switch confType {
+	case "prometheus":
+		{
```
Same here

rhsakarpos replied:
removed

xing-yang commented on pkg/api/controllers/fileops.go:
```diff
@@ -0,0 +1,193 @@
+		if toCallReloadEndpoint == true {
+			reloadResp, reloadErr := http.Post(url+reloadPath, "application/json", nil)
+			if reloadErr != nil {
+				log.Errorf("Error on reload of configuration %s", reloadErr)
```
Start the msg with lower case.

rhsakarpos replied:
updated.

xing-yang commented on pkg/api/controllers/fileops.go:
```diff
@@ -0,0 +1,193 @@
+			}
+			respBody, readBodyErr := ioutil.ReadAll(reloadResp.Body)
+			if readBodyErr != nil {
+				log.Errorf("Error on reload of configuration %s", reloadErr)
```
Start the msg with lower case.

rhsakarpos replied:
updated

leonwanghui commented on pkg/api/controllers/fileops.go:
```diff
@@ -0,0 +1,193 @@
+// Copyright (c) 2019 OpenSDS Authors All Rights Reserved.
```
Please remove `All Rights Reserved`

rhsakarpos replied:
removed

rhsakarpos commented:
> I mean can you move the code in `fileops.go` to `metrics.go`. Because downloadconf and uploadconf is subpath under metrics.

refactored, moved into metrics.go

leonwanghui commented:
Please rebase it.

rhsakarpos commented:
> Please rebase it.

Done

leonwanghui commented on pkg/api/routers/metric.go:
```diff
@@ -22,6 +22,16 @@ import (
 
+	// add router for file ops api
+	filens :=
+		beego.NewNamespace("/"+constants.APIVersion,
```
Please append it in the existing namespace.

rhsakarpos replied:
done

leonwanghui commented on pkg/api/routers/metric.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2019 OpenSDS Authors.
+// Copyright (c) 2019 OpenSDS Authors
```
Please add `.`

rhsakarpos replied:
added

leonwanghui commented:
Also need a rebase.

rhsakarpos commented on pkg/api/controllers/metrics.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2019 The OpenSDS Authors All Rights Reserved.
+// Copyright (c) 2019 The OpenSDS Authors
```
added

leonwanghui commented on pkg/api/controllers/metrics.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2019 The OpenSDS Authors All Rights Reserved.
+// Copyright (c) 2019 The OpenSDS Authors
```
Still missing here.

leonwanghui commented on pkg/api/routers/metric.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2019 OpenSDS Authors.
+// Copyright (c) 2019 OpenSDS Authors
```
Still missing here.

leonwanghui commented:
@wisererik Please take a look at this PR.
